### ISSUE TITLE
Generic value transfer primitive - master thesis

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#6029]
Problem description: _architecture and deployment of trustworthy offline value transfer_

The financial system is the heart of capitalism. In the online world we now have the Bitcoin open protocol for crypto tokens. No regulated alternative exists with sufficient consumer protection, stability, and state-based value guarantee. Central Bank Digital Currencies are being discussed by the European Central Bank, a [digital Euro](https://www.ecb.europa.eu/pub/pdf/other/Report_on_a_digital_euro~4d7268b458.en.pdf). You will build upon the [Delft University experimental deployment of a digital Euro](https://www.delta.tudelft.nl/article/delft-cafe-premieres-eemcs-blockchain-euro), turning it into a generic value transfer primitive. [Source code](https://github.com/Tribler/trustchain-superapp/blob/master/README.md)

Requirements: The value transfer primitive is required to be a cash replacement :euro: . Value transfers do not require **any** intermediary. The requirements are to support low-value transactions, offline-first (Internet is optional), and compliance with existing laws. Value transfers may be fully automated in a machine-to-machine economy. A simple scenario is that a Tesla car could use this value transfer to buy or sell battery charge automatically.

This work is ideology driven. This works supports Delft University research towards a _reformed financial sector that is more efficient, more effective in serving the wider economy and more resistant to bad behaviour of all kinds_, [Background reading on bearer instruments and central bank digital currency](https://blog.dgwbirch.com/?tag=digital-bearer-instrument).

Digital identity is required to open a bank account. Unlimited anonymous money is not compliant with money laundering and anti-terror legislation. The problems with digital identity are unsolved. You will build upon the recent work by Rowdy who deployed a self-sovereign digital identity, eWallet, QR-code scanner, and zero-knowledge proof system. The UK minister of digital infrastructure [announced their digital identity and digital trust framework](https://www.gov.uk/government/publications/the-uk-digital-identity-and-attributes-trust-framework) also the [Dutch vision on digital identity](https://www.rijksoverheid.nl/documenten/kamerstukken/2021/02/11/kamerbrief-over-visie-digitale-identiteit) by our secretary-of-state (in Dutch only). ([see also](https://www.noraonline.nl/wiki/Uitgave_en_beheer_van_digitale_bronidentiteiten))

This master thesis is part of an incremental effort ongoing for 22 years at Delft University. We are trying to make the Internet more open, free, and fair. We are building open protocols for money, trust, code and data. A generic value transfer primitive is the cornerstone of any online economy. You will be the first to create one that works and scales.
![image](https://user-images.githubusercontent.com/325224/111442759-fa137780-8708-11eb-8a72-6ee2813388b7.png)

**Comment by synctext** [comment]
Rough timeline (tentative)
- 1-2 months: read background, compile prior code, understand ID,wallet and CBDC
- 2 months: participate in design team of Digitale Bron Identiteit (Deliver a report, English thesis appendix)
- 5 months: value transfer primitive

**Comment by synctext** [comment]
Update on 30 March 2021. Many students are contributing to Superapp: [1) Machine Learning, 2) multi-sig Bitcoin, 3) liquidity pool Euro-Bitcoin 4) liquidity pool Bitcoin-Ethereum](https://github.com/Tribler/tribler/issues/5944#issuecomment-810002552)

**Comment by synctext** [comment]
Some related developments:
digital offline Euro... "The analysis confirms, by and large, our initial findings: what the public and professionals want the most from such a digital currency is privacy (43%), followed by security (18%), the ability to pay across the euro area (11%), no additional costs (9%) and offline usability (8%).", https://www.ecb.europa.eu/press/pr/date/2021/html/ecb.pr210414~ca3013c852.en.html

“States risk ceding control over many areas of finance and the economy to the private companies that dominate the digital space. Leading technology companies will entrench their monopoly positions, drawing on financial resources and superior engineering capacities. The monopoly on data held by third parties will continue to pose increasing risks of manipulation and criminal use of personal information. Privacy and the ethical use of data are key topics to be addressed by law enforcement, legislators and policy makers.”
https://www.europol.europa.eu/activities-services/main-reports/european-union-serious-and-organised-crime-threat-assessment

**Comment by synctext** [comment]
To discuss: file transfer progress bar. (https://github.com/Tribler/tribler/issues/6023#issuecomment-828205116)

**Comment by jwbambacht** [comment]
First version problem description.

[ProblemDescription.pdf](https://github.com/Tribler/tribler/files/6419483/ProblemDescription.pdf)

**Comment by synctext** [comment]
- Stopping all manual processes, [straight-through digital processing](https://en.wikipedia.org/wiki/Straight-through_processing) and achieve a unbelievable 100x drop in payment infrastructure cost.
- Do you want the current US-based monopolists or you [prefer a European-grown monopoly](https://www.ecb.europa.eu/press/pr/date/2020/html/ecb.pr200702~214c52c76b.en.html)

- Prior SSI work, generic framework for attestations: https://github.com/Tribler/tribler/issues/4461
- QR codes
- File transfer 
- DHT reliability
- Connectability
- Proxy bandwidth

**Comment by jwbambacht** [comment]
Example App Sketch

![AppOverviewNew](https://user-images.githubusercontent.com/31626076/116973427-d0181380-acbc-11eb-8e0c-e525e347072f.png)
![AppOverviewNew-2](https://user-images.githubusercontent.com/31626076/116973432-d27a6d80-acbc-11eb-937d-f89f902a780d.png)

**Comment by rwblokzijl** [comment]
I can help getting the eurtoken code to work. Do you have any way i can reach you?

**Comment by jwbambacht** [comment]
Updated version of problem description and a start on state of the art.

[ProblemDescription_02-06-2021.pdf](https://github.com/Tribler/tribler/files/6586565/ProblemDescription_02-06-2021.pdf)

**Comment by synctext** [comment]
Feel free to make it stronger, if you feel that way. You can keep it in this more "boring"/rational manner.
Within Tribler team we are interested to the point of obsession with monopolies: I. the financial system with natural banking monopolies; II. global media and news organisation with structural bias, [state-censorship](Media without censorship (CensorFree) scenarios), and fake news. III. Platform power with Google, Facebook, and Amazon with anti-competitive behaviour, winner-takes-all dynamics where competitive pressure has failed.

Opening line of problem description could be: _the current financial system and digital economy is structurally unfair to citizens. Governments have failed to protect "consumer welfare". Citizens and economic actor have no alternative to banking services, Big Tech and their anti-competitive practices [[1](https://digitalcommons.law.yale.edu/cgi/viewcontent.cgi?article=5785&context=ylj); [2](https://www.law.ox.ac.uk/business-law-blog/blog/2017/04/e-scraper-and-e-monopsony)].  This thesis contributes to a reformed financial/tech sector that is more efficient (I.), more effective in serving the wider economy (II.) and more resistant to bad behaviour of all kinds (III.)._

Our first motivating example of market failure is the WhatsApp messaging platforms. [etc etc] Whatsapp failing terms-of-service. Competitor based on openness Signal has barriers to market entry, no network effect, compete against closed protocol. Whatsapp has stickyness, like Hotel California, you can install, but never leave.

**Comment by jwbambacht** [comment]
Screenshots of current iteration of app

![Screenshot1](https://user-images.githubusercontent.com/31626076/122879234-f8c19e80-d338-11eb-8669-a2599c21487a.JPG)
![Screenshot2](https://user-images.githubusercontent.com/31626076/122879240-f95a3500-d338-11eb-8607-76d41d2ed360.JPG)
![Screenshot6](https://user-images.githubusercontent.com/31626076/122879242-f9f2cb80-d338-11eb-9b6c-56794bc0a714.JPG)
![Screenshot5](https://user-images.githubusercontent.com/31626076/122879244-fa8b6200-d338-11eb-8469-c4df23db659e.JPG)
![Screenshot3](https://user-images.githubusercontent.com/31626076/122879247-fa8b6200-d338-11eb-8994-2c701c0b0dcf.JPG)
![Screenshot4](https://user-images.githubusercontent.com/31626076/122879249-fb23f880-d338-11eb-924d-359917f5d037.JPG)

**Comment by jwbambacht** [comment]
Questions:
* conventional identity to digital identity adding/importing possible solution:
  1. show QRcode containing public key
  2. let the 3rd party (or other device) scan the QRcode
  3. conventional identity provider (or other device) creates a QRcode containing identity data encrypted using public key
  4. scan the created QRcode and decrypt using private key
  5. insert data in identity stor
* ideally an identity will always have the same public key for recoverability and consistency purposes. How to ensure that an identity will always map to the same public key? Give the user an seed phrase for instance after first initialization? What to do when the user does not know this phrase anymore? A seed phrase that is derived from the identity information can be mimicked by a malicious actor, which therefore can access the user's account. Can we expect the user is responsible for to keep it save and stored somewhere?
* change public key of device in superapp to match personal identity public key
* chats between identities not on the chain due to privacy concerns

**Comment by synctext** [comment]
The thesis work is progressing nicely. Running code is looking great!
We need to balance thesis direction: 
1. value transfer primitive
1. Technical debt and refactoring (current focus: boost architecture cleanness, GUI, polish UX) Related work "[Refactoring, Bug Fixing, and New Development Effect on Technical Debt: An Industrial Case Study](https://doi.org/10.1109/SEAA51224.2020.00068)"
1. Reliability testing infrastructure  
1. identify faults in Trustchain and IPv8, 99% reliability. Enhancing algorithms towards 99.999% reliability. (

- "let the 3rd party (or other device) scan the QRcode", lets keep this out-of-scope and keep-as-is
- ToDo recommendation: spend 1+ week on https://github.com/budtmo/docker-android
  1. running locally
  1. run Superapp
  1. multiple Superapps communicate

**Comment by jwbambacht** [comment]
The current iteration of the app. Identity and contacts/chats implemented for the biggest part. Attestations loaded from the database can be viewed as well. Adding/verifying attestations is on the to-do list. The exchange part must be implemented completely. 
<IMG src="https://user-images.githubusercontent.com/31626076/124653817-0e3cd980-de9e-11eb-9c88-d9dcb4fe79c2.gif" width=350>

Have been struggling to get docker-android going. A lot of errors and problems come up during the process. There is also a lack of documentation on how to set it up properly. I'll spend more time in the coming weeks. It was hard to spend more than a couple of hours at once due to frustration.

**Comment by synctext** [comment]
- Exact thesis direction, usability test with (25 -) 50 people?
- Usability test with 39 users done within context of Bsc thesis on [SSI with 5 students](https://github.com/Tribler/tribler/issues/6050#issuecomment-872046736), the specific report is here:
https://github.com/Tribler/tribler/files/6736098/A_Universal_Framework_for_Claim_Portability_in_Self_Sovereign_Identity_Applications.pdf
![image](https://user-images.githubusercontent.com/325224/124716634-981f8d80-df04-11eb-9675-3a64a9e0d9bd.png)
- Exact screen

**Comment by synctext** [comment]
Note, about the ID wallet. Please consider adding TUDelft+logo as an issuer of diplomas. We can easily issue course credits in a live trail. 1 Oct 2021 the European Commission is collecting prototypes from many EU teams, their focus is cross-border diploma validity, see page 27: https://northsearegion.eu/media/16680/ebsi-explained-by-daniel-du-seuil.pdf

**Comment by jwbambacht** [comment]
The current iteration of the app. Added the exchange section, but is still a bit rough and sometimes not displaying the correct information. Options to transfer money by qr or directly to contact and request added. Contacts changed slightly, including ability to remove contacts and conversations. Identity now includes complementary attributes like address, email, .etc. These attributes can at a later point be shared within a chat. They are however not 'legal' attributes of an identity. 

<IMG src="https://user-images.githubusercontent.com/31626076/126707385-795d8401-4d94-46a9-9705-00485ffcb22f.gif" width=350>

**Comment by synctext** [comment]
Feedback:
- How generic is your fabric, easy to integrate diploma attestations or Corona QR codes? https://github.com/minvws/nl-covid19-coronacheck-provider-docs/blob/main/docs/providing-events-by-digid.md

**Comment by jwbambacht** [comment]
Attached is a zip file containing a (signed) apk that can be installed as a separate trustchain application

[Wetransfer](https://we.tl/t-DpXvIkJT2r)

**Comment by jwbambacht** [comment]
Big update to the app. Many new features are implemented, including interface changes and optimizations. See the (not exhaustive) list of updates below and a gif of screenshots: 
- Many attachments can be sent between contacts in chats. 
- Identity attributes can be created, modified, and shared with other contacts using the chat
- Attestations are implemented completely based on the work of Rowdy. 
- Transactions between contacts and buy/sell from/to the online exchange portal are implemented. However, the eurotoken system is a bit buggy: 
-- often the balance is unvalidated that will get validated somewhere in the future. Sometimes manually verifying the balance or obtaining eurotokens from the online exchange portal results in a validated balance
-- eurotokens that are sent to offline contacts will not be received automatically. The message that a transaction has been sent is received by the receiver, so a feature has been implemented that displays an error when a transaction is not received properly. The sender is also able to push resend in the chat when it has not been received yet.
- A QR-code scanner that automatically recognizes any QR code within the application has been implemented. The user can therefore use the scanner instead of having to browse to the correct fragment and click the right button to initiate a scan.

An updated (signed) apk can be downloaded in a zip file [here](https://github.com/jwbambacht/trustchain-superapp/blob/66fd3d8baaafcc0c6511e3a83aba60240d9d7a6a/app/demo/debug/app-demo-debug.apk.zip). The application will be installed as a separate trustchain application.

<IMG src="https://user-images.githubusercontent.com/31626076/131406056-8d380daf-834a-40e8-abf6-18f32d63b891.gif" width=300>

**Comment by synctext** [comment]
Amazing progress this summer!
We now have the following inside the superapp:
- Self-Sovereign Identity
  - :heavy_check_mark: Key management
  - :heavy_check_mark: Identity wallet
  - :heavy_check_mark: Verifiable claims
  - :heavy_check_mark: Zero-knowledge proofs and attestations
  - :heavy_check_mark: Real-time and offline revocations
  - :heavy_check_mark: No central point of failure, robust, disaster-proof (fully decentralised)
- EuroToken
  - :heavy_check_mark: Digital wallet
  - :heavy_check_mark: Send money in real-time chats
  - :heavy_check_mark: Offline transfers between wallets
  - :heavy_check_mark: Gateway for Euro to EuroToken conversion
  - :heavy_check_mark: Existing IBAN integration (TIKKIE)
  - :heavy_check_mark: No central point of failure, robust, disaster-proof (fully decentralised)

**Comment by jwbambacht** [comment]
Pull request is merged!

**Comment by jwbambacht** [comment]
New signed alternative apk of current version of the app. The merged version also contained an error when trying to send money. ([apk](https://github.com/jwbambacht/trustchain-superapp/blob/02abb73c367d3c6e05671acd20051c3d67f35b22/app-demoNew-debug.apk))

Most of the time is spent trying to fix the errors for the pull request. Since I had no proper experience within an organization and forking this process took more time than wanted. 
Changes made to app in apk above:
* removal of word "passport" in text
* no special symbol for eurotoken but instead the euro symbol
* theme switching to include a light and more general color scheme
* some minor fixes and changes

<IMG src="https://user-images.githubusercontent.com/31626076/133391952-4bc41f3f-9a93-47ed-a692-2bbd7b0243c9.jpg" width=300>

**Comment by synctext** [comment]
Open question:
- do a field trail with merely 4 people. Take photos, document discovered issues, and fix. Compress to 3 weeks of effort.
- Leave maximum time to make a minimal viable WhatsApp killer. :partying_face: :balloon: :partying_face:

**Comment by jwbambacht** [comment]
New version attached. [(apk)](https://github.com/jwbambacht/trustchain-superapp/blob/ef5097534afc440ed55ecaec9c6d99f3d9893e82/app/demo/debug/app-demo-debug.apk)

* Green theme discarded and dark theme added
* Notifications added for any type of attachment and transactions. Clicking on notification will direct to the fragment
* Chats can be archived, muted, and even blocked. Notifications are only received for unarchived, unmuted, and unblocked chats. Blocked chats do not receive any messages at all, but will still receive transactions.
* Bottom dialog for options make sure that less unnecessary information is on the screen 
* Settings fragment that allows to choose theme, see notification status of messages and transactions and in a future version will also include information like GDPR disclaimer
* Take a photo from camera and directly sent in chat (should still be compressed though)

Font size adjustability yet to figure out how to do.

<P float="left">
<IMG src="https://user-images.githubusercontent.com/31626076/135583457-046fb404-ef89-4501-84d5-765e954812b6.jpg" width=200>
<IMG src="https://user-images.githubusercontent.com/31626076/135583451-6a731c67-3caf-4020-b3d4-dba6590ab884.jpg" width=200>
<IMG src="https://user-images.githubusercontent.com/31626076/135583459-d703837c-c73c-4f45-8df1-85cd27209c6c.jpg" width=200>
<IMG src="https://user-images.githubusercontent.com/31626076/135583460-dee91ddd-300f-4ab3-8121-70845aac523b.jpg" width=200>
</P>

**Comment by synctext** [comment]
NFC scan of passport https://github.com/Tribler/tribler/issues/2918#issuecomment-299701239

**Comment by jwbambacht** [comment]
[digital voting pass](https://github.com/digital-voting-pass/polling-station-app)

**Comment by jwbambacht** [comment]
New version of the app ([apk](https://github.com/jwbambacht/trustchain-superapp/blob/a980c49a7caa660c249423bc6f2e399d158d1384/app/demo/debug/app-demo-debug.apk.zip))

This version includes the onboarding of a passport. A passport is set to be verified when both the camera and NFC chip are used to import the identity. All details are gathered from the biometric chip, so the possibility of forging is very small. When NFC is not supported on the device the details are scanned by the camera only. Since someone can forge this themselves, the identity is set to not verified. Verified users cannot change any identity details afterwards. Unverified users can only change their given names and surname. The other details are directly gathered from the camera and are validated using the check digits on the document.
A future update will include the verification status and name of identity within a sent message so the receiver is sure that he is talking to a real identity.

**Comment by synctext** [comment]
Make your app production-ready? Robust data transfer primitive, transfer progress bar for uploading, downloading, and error handling. Translate Python into Kotlin: https://github.com/Tribler/tribler/blob/main/src/tribler-core/tribler_core/modules/remote_query_community/eva_protocol.py

**Comment by jwbambacht** [comment]
Since the last meeting have been enjoying a (well deserved) holiday and focused on writing the journal article. Had some difficulties getting started in the beginning, but finally managed to get going. It is still a work in progress and there's probably too much text in the introduction, but I will look at what to do with that at a later point in time. 

[Article_22_11_2021.pdf](https://github.com/Tribler/tribler/files/7584365/Article_22_11_2021.pdf)

**Comment by synctext** [comment]
- Excellent progress again :clap: 
- "Current messaging platforms not only violate privacy awareness of its owners but also have no proper way to enforce trust between participants."; Big Tech platforms in general have this hurtful impact on society.
- "Privacy is mostly fulfilled by the data only traversing the network without making an intermediate stop at the platforms central server", poor indirect formulation.
- "Trust is an integral part when it comes to online communication, especially with involvement of money transfer." poor reader has no idea what you mean. Dollars, Euros, Bitcoins, stable coins or CBDC. No intro sentence.

**Comment by synctext** [comment]
What is the cardinal related work? Metamask seems to come close, quite mature app and combines several paradigms (ID,wallet,DeFi,exchanges,embedded browser,dApps, etc). Architecture on Github details: https://github.com/MetaMask/metamask-mobile/blob/949368384751c1379efc83e6a1a1a7070900c4b0/architecture.svg
<img src="https://user-images.githubusercontent.com/34273858/33608345-4a2344ec-d9d5-11e7-876e-ad712179ca24.jpg" width=250>

**Comment by jwbambacht** [comment]
Updated version of the app [(zip)](https://filebin.net/6xnv06stvgjxl4cr).

* Mainly worked on implementing the EVA protocol into the TrustChain application, which is working as expected after some necessary python-Kotlin code translations. The PeerChat community now applies the EVA protocol by default for the transfer of attachments. Currently set the window size of the protocol to 64, but I need a thorough performance analysis to find the optimal window size (if time would allow it). It may be interesting to see if that a single optimal value can be found for different connection compositions as well: wifi-to-wifi, wifi-to-4g, and 4g-to-4g. This would also be a contribution to the paper. I expected the transfer speeds to be a little bit faster. Hopefully finding an optimal window size (and possibly tuning the other parameters as well) help improve the performance of the protocol.
* In a chat within ConfIDapp the file size and downloading progress can be tracked to get more insight into the download process. Since IPv8 only supports one concurrent transfer other transfers will be scheduled to be downloaded after the current file has been transferred. Currently, only photos can be transferred, but I will implement the transfer of any(?) filetype as well.
* Within a chat it is now recognized when the identity of the contact has changed. The identity information that has changed will be displayed in the chat when the identity is determined initially and when it has changed during a conversation. The change will only be recognized when the contact sends a message to you, due to privacy concerns.
* Some minor changes to the app and have been working on a proper friend request scheme that includes identity information for trust enforcement.

No progress has been made with the paper since implementing the EVA protocol took quite some time. For the next sprint, I'm planning to spend about half of the time to the paper.

**Comment by synctext** [comment]
- Murphy effect: py-ipv8 code accidentally deleted, demo going wrong :hammer: :face_with_head_bandage: 
- Positive news: EVA worked!
- Main thesis elements completed, now only polish + :writing_hand: remains 
- Window size [scientific background material](https://archive.geant.org/projects/gn3/geant/services/edupert/Documents/tcp-performance_AlexGall.pdf). Make a few graphs for thesis.

**Comment by jwbambacht** [comment]
New version of app that has an updated EVA protocol for a massive increase in transfer speed.

 [https://filebin.net/kijjf087lr6u43j2](https://filebin.net/kijjf087lr6u43j2)

**Comment by jwbambacht** [comment]
New sprint, new apk [https://filebin.net/8320ndukryz5zu3w](https://filebin.net/8320ndukryz5zu3w), updated article [Article_4_1_2022.pdf](https://github.com/Tribler/tribler/files/7810542/Article_4_1_2022.pdf), and poster for next week.

* Created a poster for next week. Feedback is appreciated 😄.
* Primarily focused on improving the EVA protocol, which I think was successful. I'm not sure what the transfer speed of the first version of the Kotlin EVA protocol was, but the updated version is much faster and reached about 215kB/s on average.
* Created an EVA Test class and debug mode for the EVA protocol that logs a file on the phone storage. Executed an experimental analysis, processed the data in Python, and started reporting it in the article. Is it also interesting to include the transfer speed of PeerChat from before using the EVA protocol as this was extremely slow and unpredictable?
* App now contains a media gallery to browse and view images and files, and even has the possibility to zoom images.
* App can now send any file of the following mimeTypes: application/*, video/*, text/plain. Only images can be opened within the app.
* App can now download images and files to phone storage.
* App has an improved location view with Google Maps integration. To send a location, the user can choose its current location or drop a pin on the map. A received location can now be viewed within the app.

**Comment by synctext** [comment]
- Great first draft of poster. Internet-deployed: https://play.google.com/store/apps/details?id=nl.tudelft.trustchain
-

**Comment by jwbambacht** [comment]
New update to the article.
[Article_21_01_22.pdf](https://github.com/Tribler/tribler/files/7911575/Article_21_01_22.pdf)

**Comment by synctext** [comment]
:+1: 
- Abstract is now dominated by negativity, please discuss your solution at length.
- Introduction is missing remarks like 'these organisation have proven to be borderline incompetent at protecting all this hoard of private information'. Commercial entities have an incentive to minimise spending on cybersecurity. https://www.zdnet.com/article/the-biggest-data-breaches-of-2021/

**Comment by jwbambacht** [comment]
[digid costs](https://logius.nl/onze-organisatie/zakendoen-met-logius/doorbelasting)

**Comment by jwbambacht** [comment]
Thesis in the correct format. [Thesis_3_2_22.pdf](https://github.com/Tribler/tribler/files/7992882/Thesis_3_2_22.pdf)

**Comment by synctext** [comment]
Solid progress! Getting ready to pick a defence date. Quick scan only:
- please also [submit here](https://arxiv.org/help/submit) before your defense and please [list me as a co-author](https://arxiv.org/a/pouwelse_j_1.html)
- Figures can be further polished; try to use a single graphical style
- Measurement results are bit too small to understand

**Comment by synctext** [comment]
- Polish phase beginning!
- 12MByte thesis, compress the picture with 50% perhaps?
- Small textual polish required: "Ownership of identities is generally with governments."
-

**Comment by synctext** [comment]
> Fig. 13: Evaluation of the performance of a transfer of 250MB

This is very poor performance. 90% [or something is lost](https://www.mdpi.com/2071-1050/14/2/829/pdf). These GHz devices and their 4G and/or Wifi5 transfers can do more.

**Comment by jwbambacht** [comment]
Updated draft [Thesis_17_02_22.pdf](https://github.com/Tribler/tribler/files/8087513/Thesis_17_02_22.pdf)

**Comment by synctext** [comment]
Solid tuning! Getting ready for defence.
<IMG src=https://user-images.githubusercontent.com/325224/154455589-18b1c79b-c147-4f23-945b-8eddae37bcef.png width=300>
The maximum parameter settings are big windo and big block, expressed in KByte: 128 x 1200 / 1024 = 150 KByte of unacknowledged data as the fundamental limit to data transfer performance.

**Comment by jwbambacht** [comment]
[EVA kotlin tests](https://github.com/Tribler/kotlin-ipv8/tree/master/ipv8/src/test/java/nl/tudelft/ipv8/messaging/eva)

**Comment by jwbambacht** [comment]
[Defence_Presentation v1.pdf](https://github.com/Tribler/tribler/files/8142085/Defence_Presentation.v1.pdf)

**Comment by synctext** [comment]
Quick comments on .PPT
- slide with "Problem Description" is scientific terms and with scientific depth.

**Comment by synctext** [comment]
Thesis complete :clap: :partying_face: :clap: 
https://repository.tudelft.nl/islandora/object/uuid%3A3ad68dbd-3444-4e01-94a2-d28044b0ba3f
"Web3: A Decentralized Societal Infrastructure for Identity, Trust, Money, and Data"

**Comment by henrysmit420** [comment]
Mine bitcoin for free here https://www.btcs.love/invite/4nc58
